### ISSUE TITLE
rich rule masquerade needs to allow masqueraded packets on the zone

### ISSUE THREAD

**Issue by hwoarang** [firewalld/firewalld#80]
Hi,

I am trying to use the rich rule masquerade functionality to only allow a specific subnet to see the inet via masquerade. `--add-masquerade` is not suitable because it does not do source based filtering so I use the following rich rule in the external zone (the one connected to inet)

`rule family=ipv4 source address=192.168.123.0/24 masquerade`

This seems to generate the following rule in the nat table

`-A POST_external_allow -s 192.168.123.0/24 ! -o lo -j MASQUERADE`

but there is no matching rule in the filter table. Here is a diff when i run `firewall-cmd --zone=external --add-masquerade` to the already running instance

```
 -A FWDO_external -j FWDO_external_deny
 -A FWDO_external -j FWDO_external_allow
 -A FWDO_external_allow -d 192.168.123.0/24 -m conntrack --ctstate NEW -j ACCEPT
+-A FWDO_external_allow -j ACCEPT
 -A FWDO_internal -j FWDO_internal_log
 -A FWDO_internal -j FWDO_internal_deny
 -A FWDO_internal -j FWDO_internal_allow
```

As you see, when you set `--add-masquerade` to the zone, a suitable ACCEPT target is created in the `FWDO_external_allow` zone which actually allows the forwarded packets.

I believe the previously mentioned rich rule should do the same thing since you want masquerade for the specified <source network> so a matching ACCEPT target should be created.

Using the following direct rule workarounds the problem and makes masquerading for the said <source network> to work as expected

`--add-rule ipv4 filter FWDO_external_allow 1 -s 192.168.123.0/24 -j ACCEPT`

What are your thoughts?

**Comment by t-woerner** [comment]
The general ACCEPT rule is accepting everything. This is not correct and should not be needed. Is the traffic for 192.168.123.0/24 handled by the interface that is bound to the external interface? If you want to masquerade only 192.168.123.0/24, then bind this source to the external interface instead.

**Comment by olebowle** [comment]
Hi,

I basically have the same question as hwoarang. I only want to masquerade 100.64.0.0/10. Currently it looks like this:
```
cpe_priv
  interfaces:
  sources: 100.64.0.0/10
  services: dhcp dns
  ports:
  masquerade: no
  forward-ports:
  icmp-blocks:
  rich rules:

external (active)
  interfaces: enp11s0
  sources:
  services: http https
  ports:
  masquerade: yes
  forward-ports:
  icmp-blocks:
  rich rules:
```
So masquerading everything that leaves enp11s0 and allowing 100.64.0.0/10 to do dhcp and dns. If I understood you correctly, you proposed something like this:
```
external (active)
  interfaces: enp11s0
  sources: 100.64.0.0/10
  services: dhcp dns http https
  ports:
  masquerade: yes
  forward-ports:
  icmp-blocks:
  rich rules:
```
However, this would mean that everyone communicating through enp11s0 would be able to do dhcp and dns, isn't it? Is the following config the correct and most elegant solution from your point of view?
```
external (active)
  interfaces: enp11s0
  sources: 100.64.0.0/10
  services: http https
  ports:
  masquerade: yes
  forward-ports:
  icmp-blocks:
  rich rules:
	rule family="ipv4" source address="100.64.0.0/10" service name="dhcp" accept
	rule family="ipv4" source address="100.64.0.0/10" service name="dns" accept
```
Thanks!

**Comment by phemmer** [comment]
Ditto on this issue. I have multiple subnets connected to a machine, and I only want to masquerade packets coming from one of them.

> If you want to masquerade only 192.168.123.0/24, then bind this source to the external zone instead and make sure that masquerading is enabled in that zone.

This does not address the issue. When you enable masquerade for a zone, it creates a rule enabling masquerade based on the destination, not the source. I want to masquerade packets coming from the zone, not going to it.

**Comment by erig0** [comment]
@hwoarang, AFAICS, there is a rule created in the FWDO_<zone>_allow chain, but it's mistakenly using destination filtering instead of source.

https://github.com/firewalld/firewalld/blob/0e330d6ed3a7f90066832e60ed459cc90c475072/src/firewall/core/fw_zone.py#L1887-L1891

    # firewall-cmd --add-rich-rule "rule family=ipv4 source address=192.168.123.0/24 masquerade"

    # iptables  -t filter  -nvL
    Chain FWDO_public_allow (1 references)
    pkts bytes target     prot opt in     out     source               destination
       0     0 ACCEPT     all  --  *      *       0.0.0.0/0            192.168.123.0/24     ctstate NEW

**Comment by erig0** [comment]
@hwoarang, @phemmer, @olebowle,
I believe pr #329 fixes this. I would appreciate it if one of you could give it a try.

**Comment by erig0** [comment]
Fixed by  c80fd36, aka pr #352.
